### DISCUSSION TITLE
wazuh-agent.exe stop driectly - faulting module sysinfo.dll

### DISCUSSION THREAD

**Discussion by str4int** [wazuh/wazuh#20382] [Q&A]
Dear Community,

I've question about wazuh agent v4.6.0 on **virutal** windows 10 entrerprise 22H2.
I install wazuh form command line and by msi but same behavior.
When the service _WazuhSvc_ is started it stop righ after.

I have this event from windows:
    _Faulting application path: C:\Program Files (x86)\ossec-agent\wazuh-agent.exe
    Faulting module path: C:\Program Files (x86)\ossec-agent\sysinfo.dll_

strangely into ossec.log i dont find any Error or Warning.
sysinfo seems to be verify properly:
    _- wazuh-agent[0] cryptography.c:541 at verify_hash_and_pe_signature(): DEBUG: PE signature verification succeeded for C:\Program 
    Files (x86)\ossec-agent\sysinfo.dll
    - wazuh-agent[0] dll_load_notify.c:101 at dll_notification(): DEBUG: The file 'C:\Program Files (x86)\ossec-agent\sysinfo.dll' is signed 
    and its signature is valid._

FYI, i don't have any issue on server.
Do you know what it could be ?

Best,

**Comment by str4int** [comment]
Dear Community,

Seems that the faulty module is the "paquages" module into Syscollector.

By disabling it the agent work "properly"
ossec.conf:

    
    <wodle name="syscollector">
      <disabled>no</disabled>
      <interval>1h</interval>
      <scan_on_start>yes</scan_on_start>
      <hardware>yes</hardware>
      <os>yes</os>
      <network>yes</network>
      <packages>no</packages>
      <ports all="no">yes</ports>
      <processes>yes</processes>
  
      
      <synchronization>
        <max_eps>10</max_eps>
      </synchronization>
    </wodle>


Do you know what's causing it?
best,

**Reply by pereyra-m** [reply]
Hello @St4nZ !
Thank you for reporting this issue.

When there is a problem with **sysinfo**, we follow this same procedure:
https://github.com/wazuh/wazuh/issues/20344#issuecomment-1820953694

You can use this debugging tool to run the same code the agent does
[testSysinfo.zip](https://github.com/wazuh/wazuh/files/13553700/testSysinfo.zip)

Then, you run one type of scan at a time until you confirm the issue (in this case, you've already narrowed the problem to packages). For example:

` ./sysinfo_test_tool --packages`

If the problem is with packages, then please upload the output of the tool and the full software inventory of the agent so we can reproduce the issue.

Regards.

**Reply by str4int** [reply]
Hi @pereyra-m ,
Thank you for your message.
Unfortunatelly same error when running _./sysinfo_test_tool --packages_ without output.
Best,

[apps.txt](https://github.com/wazuh/wazuh/files/13560289/apps.txt)
![sysinfo_error](https://github.com/wazuh/wazuh/assets/72645609/1d90c9d9-eafd-4e8d-8c7e-e33c03b4690a)

**Reply by pereyra-m** [reply]
Thank you for the information provided @St4nz!

I've seen the apps you shared and I don't find any of them with an unusual name or character that might cause a parsing issue.
According to the application crash you've shared, an operation of a JSON object with an external library is causing the problem

```
0:001> ln sysinfo + 0x17dba

 [\usr\lib\gcc\i686-w64-mingw32\10-posix\include\c++\bits\std_function.h @ 283] (73017d68)   sysinfo!_M_manager+0x52   |  (73017dc9)   sysinfo!unescape<std::__cxx11::basic_string<char> >

0:001> u 0x73000000 + 0x17dba
sysinfo!_M_manager+0x52 [\usr\lib\gcc\i686-w64-mingw32\10-posix\include\c++\bits\std_function.h @ 283]:
73017dba e884030000      call    sysinfo!_M_manager (73018143)
73017dbf b800000000      mov     eax,0
73017dc4 8b5dfc          mov     ebx,dword ptr [ebp-4]
73017dc7 c9              leave
73017dc8 c3              ret
sysinfo!unescape<std::__cxx11::basic_string<char> > [\home\vagrant\wazuh_win_2\wazuh\src\external\nlohmann\json.hpp @ 2989]:
73017dc9 55              push    ebp
73017dca 89e5            mov     ebp,esp
73017dcc 53              push    ebx
```

I'm unable to reproduce it locally, could you run this new compilation of the debugging tool again? We'll also enable the crash dumps to collect more precise data

- First, run this registry entry to enable the dumps [enable_wazuh_dumps_test_tool.zip](https://github.com/wazuh/wazuh/files/13573922/enable_wazuh_dumps_test_tool.zip)
- Then, run again the test tool in the same way as before  [sysinfo-4.7.0.zip](https://github.com/wazuh/wazuh/files/13573973/sysinfo-4.7.0.zip)
- Now, after the crash, a dump file will be available in `C:\CrashDumps`. Please upload it here. We will analyze them with the proper symbols ([sysinfo_symbols.zip](https://github.com/wazuh/wazuh/files/13574008/sysinfo_symbols.zip))

Regards.

**Reply by str4int** [reply]
Thanks for your input.

Unfortunatelly, for professionnal reason, i can't share this dump here.
I will try to reproduce the issue on a lab environnment and provide you the files.

Many thanks for your support so far.
Best,

**Reply by pereyra-m** [reply]
Hi again @St4nZ !
I understand, thanks to you.

We'll be waiting for your update.
Regards.

**Reply by weeix** [reply]
@pereyra-m I also encountered this very issue in v4.8.1. Here is the dump file:

(deleted as the issue has been fixed in v4.9.1)

**Reply by pereyra-m** [reply]
Thank you @weeix for the core dump.
This information is essential to analyze the bug and fix it.

I'll inform the @wazuh/devel-agent team so they can continue the research.

**Comment by JBent90914** [comment]
From MSDN, both gmtime() and localtime() can (and likely will) return NULL.

https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/gmtime-gmtime32-gmtime64?view=msvc-170

https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/localtime-localtime32-localtime64?view=msvc-170

If an installed package has an install date outside those ranges, the Utils::getTimestamp() call in timeHelper.h will fail.

Patch below.

```
diff -ruN wazuh-4.8.2/src/shared_modules/utils/timeHelper.h new/src/shared_modules/utils/timeHelper.h
--- wazuh-4.8.2/src/shared_modules/utils/timeHelper.h   2024-08-19 12:13:17.000000000 -0400
+++ new/src/shared_modules/utils/timeHelper.h   2024-08-27 00:02:36.065955978 -0400
@@ -29,6 +29,10 @@
         std::stringstream ss;
         // gmtime: result expressed as a UTC time
         tm* localTime {utc ? gmtime(&time) : localtime(&time)};
+
+        // Return "invalid" timestamp, if the input is invalid.
+        if (localTime == NULL) { return("1970/01/01 00:00:00"); }
+
         // Final timestamp: "YYYY/MM/DD hh:mm:ss"
         // Date
         ss << std::setfill('0') << std::setw(4) << std::to_string(localTime->tm_year + 1900);
```

**Reply by weeix** [reply]
I replaced the original sysinfo.dll with a custom-built version using this patch. It worked perfectly, allowing me to enable package scanning without crashing the agent. Thanks!
